### ISSUE TITLE
[Feature request] File upload max size

### ISSUE THREAD

**Issue by KaKi87** [PrivateBin/Directory#120]
Hi,

As many instances limit file sharing to only a few megabytes, it would be nice to display that max size in the table instead of just a check mark.

Thanks

**Comment by elrido** [comment]
That would be helpful, yes. Unfortunately that limit is currently not exposed to the client/webUI. It was [once discussed but no progress has been made](https://github.com/PrivateBin/PrivateBin/issues/95) on it yet.

Currently the only way to detect this is by trial and error (sending larger and larger pastes until the size error occurs), so that is not something I'd want to do on listed instances once per day or so. Even when taking care we delete the test pastes immediately, it is a quite aggressive test compared to sending a single GET request once per day and sending HEAD requests every 15 min.

**Comment by KaKi87** [comment]
You could only do it when an instance is added or an existing instance's version number is changed (that's the likeliest time at which configuration may be changed), or if you'd rather stick to a simple cron then once a month.

And you could do it with only a few ranges ranges, e.g. :

- Start with 5 Mb and show max file size as `< 5 Mb` if failing ;
- Continue with 10 Mb and show max file size as `< 10 Mb` is failing ;
- Jump to 100 Mb and show max file size as `< 100 Mb` is failing ;
- Stop at 500 Mb and show max file size as `< 500 Mb` if failing or `> 500 Mb` is succeeding.

I think this would satisfy most use cases.

**Comment by elrido** [comment]
The point of this solution is to be self curating, so doing anything only once is not in scope (and require quite a fundamental design change). Doing such an aggressive scan (bordering on an attack) is bad practice IMHO and will only lead to less instances wanting to participate in this list.

Finally it is very unlikely we would even find any instances with more than 100 MiB, as the design of PrivateBin, where every message must in full be transmitted in a single POST request and gets loaded into RAM by the PHP process at least twice (when decoding the JSON and later to validate the base64 encoded payload actually contains something that resembles encrypted data) makes this very expensive in terms of resources. An instance accepting 1 GiB pastes would require POST limits above 1GiB on webserver and in PHP and a PHP memory limit of 2-3 GiB per process and so you'd need to get a server with 16 GiB or RAM or more to allow for the typical 5 parallel PHP workers. Sending 6 large requests, slowly, to this server would already block uploads for all visitors.

The typical PHP memory limits will be around 25 MiB, so the default 10 MiB upload limit suits that. This allows for dozens of parallel workers even on a small instance. The privatebin.net demo runs on 512 MiB RAM, shared also with this application and the static website. For a single photo, a PDF or office document this should be plenty as these all are usually only a few MiBs in size.

We can keep this open and I will certainly consider revisiting this feature if and when we publish a size limit that we can easily scrape out of the HTML or request via an API.

**Comment by KaKi87** [comment]
> every message must in full be transmitted in a single POST request and gets loaded into RAM by the PHP process at least twice [...] makes this very expensive in terms of resources

Oh, I didn't know that.

Well, I was trying to find an instance on which I could transfer 30-80 MB videos 😅

Is PrivateBin therefore unsuitable for 
E2E encrypting & transferring large files ? What would you then recommend for that purpose ?

**Comment by elrido** [comment]
PrivateBin is primarily intended to exchange texts and optionally (disabled by default) a single document alongside. Small videos are possible, but it is certainly not designed with sharing large videos.

I don't know of E2E software in that area, but personally would use NextCloud for large files or when sharing collections of many files - while not E2E encrypted, it is similarly self-hosted, supports/requires creation of user accounts and each user account optionally gets their own (server side, not E2E) key to encrypt files at rest, so the server can only decrypt the file when the user has authenticated or with a shared key from a link created for sharing the file(s).

**Comment by KaKi87** [comment]
Yeah, I'm looking for something that works when my self-hosted stuff isn't, and since it's third-party, I wouldn't want something that's not E2E encrypted.

So, short of a better alternative, I'm gonna make this scan myself using Puppeeteer, I'll share the results here if you're interested.

Thanks anyway
